pkg/executor/infoschema_reader.go

```diff
@@ -112,31 +114,42 @@ func (e *memtableRetriever) retrieve(ctx context.Context, sctx sessionctx.Contex
 	if !e.initialized {
 		is := sctx.GetInfoSchema().(infoschema.InfoSchema)
 		e.is = is
-		dbs := is.AllSchemaNames()
-		slices.SortFunc(dbs, func(a, b model.CIStr) int {
-			return strings.Compare(a.L, b.L)
-		})
+
+		var getAllSchemas = func() []model.CIStr {
+			dbs := is.AllSchemaNames()
+			slices.SortFunc(dbs, func(a, b model.CIStr) int {
+				return strings.Compare(a.L, b.L)
+			})
+			return dbs
+		}
+
 		var err error
 		switch e.table.Name.O {
 		case infoschema.TableSchemata:
+			dbs := getAllSchemas()
```

Comment:
why not eval them once before switch case

Reply:
Because some of the cases don't need this.